src/Runtime/jni/src/com/ibm/onnxmlir/OMModel.java

```diff
@@ -1,58 +1,81 @@
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.Comparator;
+import java.util.Enumeration;
+import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
+import java.util.logging.Logger;
 
 public class OMModel {
-    static String libname = "libmodel.so";
+    private static final Logger logger = Logger.getLogger(OMModel.class.getName());
 
     static {
-        File jar;
-        JarFile jf;
-        String jarDir = null;
-        String libPath = null;
+	OMLogger.initLogger(logger);
+
         try {
-            // Get path name of jar
-            jar = new File(OMModel.class.getProtectionDomain()
-                                        .getCodeSource()
-                                        .getLocation().toURI());
-            jarDir = jar.getParentFile().getAbsolutePath();
-            libPath = jarDir + "/" + libname;
-
-            // Open jar file to read and check libname inside jar.
-            // If IOException thrown, load .so from where .jar is.
-            //
-            // Checking whether DynEntryPoint.class.getResourceAsStream returns null
-            // does NOT work. Because it checks whether the resource is
-            // available on the classpath, not only just inside the jar file.
-            jf = new JarFile(jar);
-            if (jf.getEntry(libname) != null) {
-                File libFile = new File(libPath);
-                // Copy .so to where jar is
-                Files.copy(jf.getInputStream(jf.getEntry(libname)),
-                        libFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
-                // z/OS USS requires "x" permission bit
-                libFile.setExecutable(true, false);
-                // Load the temporary .so copy
-                System.load(libPath);
-                // POSIX can unlink file after loading
-                libFile.delete();
-            }
-            else {
-                // Throw subclass of IOException
-                throw new FileNotFoundException(".so not found inside jar");
-            }
-        } catch (URISyntaxException e) {
-            // Failed to find jar path, assume the .so is in cwd
-            System.load(libname);
-        } catch (IOException e) {
-            // .so not found in jar, assume it's where jar is
-            System.load(libPath);
+            // Get path of jar
+	    Path jar = Paths.get(OMModel.class.getProtectionDomain()
+                                              .getCodeSource()
+                                              .getLocation().toURI());
+	    Path jarDir = jar.getParent();
+
+	    // All .so files found will be copied under the ./native subdirectory
+	    // where the jar file resides.
+	    Path tmpDir = Files.createTempDirectory(jarDir, "native_");
+
+            // Open jar file to load all .so libs inside. The .so is loaded
+            // by OS so it must be extracted to the file system.
+            JarFile jf = new JarFile(jar.toFile());
```

Comment:
nit, there appears to be a minor indentation error here.

Reply:
No problem in the file. Github is somehow displaying the diff wrong.